File: mpi/traffic/traffic/C-SER/traffic.c
#include <stdlib.h>
#include <stdio.h>

#include "traffic.h"

#define NCELL 100000

int main(int argc, char **argv)
{
  int *oldroad, *newroad;

  int i, iter, nmove, ncars;
  int maxiter, printfreq; 

  float density; 

  double tstart, tstop;

  oldroad = (int *) malloc((NCELL+2)*sizeof(int));
  newroad = (int *) malloc((NCELL+2)*sizeof(int));

  maxiter = 200000000/NCELL; 
  printfreq = maxiter/10; 

  // Set target density of cars

  density = 0.52;

  printf("Length of road is %d\n", NCELL);
  printf("Number of iterations is %d \n", maxiter);
  printf("Target density of cars is %f \n", density);

  // Initialise road accordingly using random number generator

  printf("Initialising road ...\n");
  
  ncars = initroad(&oldroad[1], NCELL, density, SEED);

  printf("...done\n");

  printf("Actual density of cars is %f\n\n", (float) ncars / (float) NCELL);

  tstart = gettime();

  for (iter=1; iter<=maxiter; iter++)
    { 
      updatebcs(oldroad, NCELL);

      // Apply CA rules to all cells

      nmove = updateroad(newroad, oldroad, NCELL);
      
      // Copy new to old array

      for (i=1; i<=NCELL; i++)
	{
	  oldroad[i] = newroad[i]; 
	}

      if (iter%printfreq == 0)
	{
	  printf("At iteration %d average velocity is %f \n",
		 iter, (float) nmove / (float) ncars);
	} 
    }

  tstop = gettime();

  free(oldroad);
  free(newroad);
  
  printf("\nFinished\n");
  printf("\nTime taken was  %f seconds\n", tstop-tstart);
  printf("Update rate was %f MCOPs\n\n", \
	 1.e-6*((double) NCELL)*((double) maxiter)/(tstop-tstart));
}

File: mpi/traffic/traffic/F-SER/traffic.f90
program traffic

  use trafficlib

  implicit none

  integer, parameter :: NCELL = 100000
  integer, parameter :: maxiter = 200000000/NCELL
  integer, parameter :: printfreq = maxiter/10

  integer :: i, iter, nmove, ncars
  real    :: density

  integer, allocatable, dimension(:) :: newroad, oldroad

  double precision :: tstart, tstop

! Set target density of cars

  density = 0.52

  write(*,*) 'Length of road is ', NCELL
  write(*,*) 'Number of iterations is ', maxiter
  write(*,*) 'Target density of cars is ', density

! Allocate arrays

  allocate(newroad(0:NCELL+1))
  allocate(oldroad(0:NCELL+1))

! Initialise road accordingly using random number generator

  write(*,*) 'Initialising ...'

  ncars = initroad(oldroad(1), NCELL, density, seed)

  write(*,*) '... done'

  write(*,*) 'Actual density of cars is ', float(ncars)/float(NCELL)
  write(*,*)

  tstart = gettime()

  do iter = 1, maxiter

     call updatebcs(oldroad, NCELL)

     nmove = updateroad(newroad, oldroad, NCELL)

! Copy new to old array

     do i = 1, NCELL

        oldroad(i) = newroad(i)

     end do

     if (mod(iter, printfreq) .eq. 0) then

        write(*,*) 'At iteration ', iter, ' average velocity is ', &
             float(nmove)/float(ncars)
     end if

  end do

  tstop = gettime()

  deallocate(newroad)
  deallocate(oldroad)

  write(*,*)
  write(*,*) 'Finished'
  write(*,*)
  write(*,*) 'Time taken was  ', tstop - tstart, ' seconds'
  write(*,*) 'Update rate was ', &
              1.d-6*float(NCELL)*float(maxiter)/(tstop-tstart), ' MCOPs'
  write(*,*)

end program traffic
